### ISSUE TITLE
[HOLD for payment 2023-08-28] [$1000] Invalid Navigation from Bank Connect flow on re-fresh.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#24237]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open main settings & navigate to the Bank connect page.
2. Refresh on this page.
3. Navigate back till the end.
4. Observe there's no navigation happened for few settings pages like (main navigation, workspace pages, etc..)
## Expected Result: 
Navigation should happen for every screen we navigated before refresh.

## Actual Result: 
Some navigation screens are skipped when clicking back repeatedly.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.51-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/87e1e7c8-c39b-4955-baf0-e5680b10099e

https://github.com/Expensify/App/assets/43996225/e56ce58e-9d64-4c47-975a-6ebe29c0993e

**Expensify/Expensify Issue URL:**
**Issue reported by:** @b4s36t4
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690926833777639

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0108f6cd09469076a9</li>
        <li>Upwork Job ID: 1689075378436157440</li>
        <li>Last Price Increase: 2023-08-09</li>
<li>Automatic offers: </li>
<ul>
<li>b4s36t4 | Reporter | 26072343</li>
</ul></li>
    </ul>
</details>

**Comment by ygshbht** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When navigating the Bank Connect flow, if a user refreshes while on the Bank connect page and attempts to navigate backward, certain settings pages (like main navigation, workspace pages, etc.) are not navigated to. Instead, the user experiences unexpected or no navigation for those specific screens.

### What is the root cause of that problem?
The root cause is that the Navigation.goBack() function, used within the Bank Connect flow, does not provide a fallback route parameter. As a result, it defaults to navigating to the HOME page when no prior route exists in the navigation stack. This behavior leads to unexpected navigation outcomes after a refresh, particularly when the previous routes aren't retained in the navigation history.
https://github.com/Expensify/App/blob/2ce4328401ec71f9ec0dc1304e0c5be261ae5628/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L278-L322
### What changes do you think we should make in order to solve the problem?
To address the described issue, we need to ensure that when navigating back, there's a specified fallback route. This can be achieved with the following steps:

1. Modify the `navigateToBankAccountRoute` function to accept a `backTo` parameter that provides the fallback route.
https://github.com/Expensify/App/blob/2ce4328401ec71f9ec0dc1304e0c5be261ae5628/src/components/ConnectBankAccountButton.js#L39
https://github.com/Expensify/App/blob/2ce4328401ec71f9ec0dc1304e0c5be261ae5628/src/libs/actions/ReimbursementAccount/navigation.js#L21-L23
2. Adjust the `getBankAccountRoute` function to account for the backTo parameter like `getMoneyRequestCurrencyRoute`
https://github.com/Expensify/App/blob/2ce4328401ec71f9ec0dc1304e0c5be261ae5628/src/ROUTES.js#L21
https://github.com/Expensify/App/blob/2ce4328401ec71f9ec0dc1304e0c5be261ae5628/src/ROUTES.js#L97
3. When invoking the `ReimbursementAccount.navigateToBankAccountRoute `function in `ConnectBankAccountButton`, pass the desired/ current active route as the `backTo` parameter.
4. In the `ReimbursementAccountPage`, while executing the `goBack` logic, retrieve the fallback route from the route params and navigate accordingly.

This is similar to how IOUCurrencySelection handles the issue: 
https://github.com/Expensify/App/blob/1c8f9353f1d3d4a14c0e9c23a5f5299aec36f17f/src/pages/iou/steps/NewRequestAmountPage.js#L143-L147

https://github.com/Expensify/App/blob/1c8f9353f1d3d4a14c0e9c23a5f5299aec36f17f/src/ROUTES.js#L97

https://github.com/Expensify/App/blob/1c8f9353f1d3d4a14c0e9c23a5f5299aec36f17f/src/pages/iou/IOUCurrencySelection.js#L83-L96

**Comment by burczu** [comment]
The proposal from @ygshbht makes sense to me as it's quite common approach in web development and we already use it in the app.

**Comment by sakluger** [comment]
Thanks @burczu! Can you please add the following comment along with the proposal you select?

> 🎀 👀 🎀 C+ reviewed

After leaving that in a comment, an Expensify engineer will automatically be assigned to review that proposal as well.

**Comment by burczu** [comment]
@sakluger done!

**Comment by b4s36t4** [comment]
@burczu i think you have to send a new comment instead of editing so that melvin can pick that.

**Comment by burczu** [comment]
@b4s36t4 Yeah, you may be right - let's see ;)

🎀 👀 🎀 C+ reviewed

**Comment by ygshbht** [comment]
Thanks. I'll have the PR ready in a day

**Comment by ygshbht** [comment]
@burczu 
The [PR](https://github.com/Expensify/App/pull/24480) is ready for review!

**Comment by ygshbht** [comment]
Hi @yuwenmemon, shoud bonus apply here? This was assigned on Friday. I created the PR within a day. Saturday/ Sunday was weekend and @burczu was OOO Monday/ Tuesday. The pull request was merged on Wednesday. Thank you very much!

**Comment by burczu** [comment]
I can confirm - I was OOO and was able to review and approve the PR on Wednesday, otherwise I would have been able to do it on Monday.

**Comment by dylanexpensify** [comment]
@yuwenmemon @sakluger do you think this will solve [this issue](https://github.com/Expensify/App/issues/24635) regarding back button navigation during bank connect flow?

**Comment by dylanexpensify** [comment]
@yuwenmemon @sakluger bumping [this](https://github.com/Expensify/App/issues/24237#issuecomment-1686303370) 🙇‍♂️

**Comment by sakluger** [comment]
I think it's fair to apply the bonus since @burczu was OOO and didn't have enough time to reassign after the issue was assigned to @ygshbht.

@dylanexpensify I think this might be the same issue as yours. I'm a little confused though - the [linked PR](https://github.com/Expensify/App/pull/24480) shows that it was deployed to staging twice but never to production. @yuwenmemon do you know what's going on there?

**Comment by yuwenmemon** [comment]
It's on prod, the deploy script can be flakey IIRC.

**Comment by ygshbht** [comment]
@sakluger Accepted!

**Comment by burczu** [comment]
@sakluger yes, I'm from Callstack

**Comment by burczu** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
 
* [x]  [@burczu] The PR that introduced the bug has been identified. Link to the PR: n/a
* [x]  [@burczu] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: n/a
* [x]  [@burczu] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: n/a
* [x]  [@burczu] Determine if we should create a regression test for this bug.
* [x]  [@burczu] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
* [ ]  [@sakluger] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by sakluger** [comment]
@burczu thanks! For the checklist, just to confirm, you're saying that this was not a regression from another PR, right? And also, you're saying that we should not add regression steps for this?

I paid @ygshbht so as soon as we confirm the checklist/regression details, we can close this issue.

**Comment by burczu** [comment]
@sakluger Yes, correct.

**Comment by sakluger** [comment]
Great, thanks!
